### PR TITLE
test: NodeClockContext follow-ups

### PR DESCRIPTION
**Pull request by seduless**
Follow-up to #34858

Updates remaining `SetMockTime` call sites that are clean, mechanical swaps fitting the spirit of the original PR (see: https://github.com/bitcoin/bitcoin/pull/34858#pullrequestreview-4031647119 and https://github.com/bitcoin/bitcoin/pull/34858#issuecomment-4221757881). Further updates to `SetMockTime` are more complex and deserve separate, isolated PRs.

The default constructor for `NodeClockContext` increments to the next tick, which is a defensive measure to prevent time going backwards on construction. This has caused some confusion (see thread: https://github.com/bitcoin/bitcoin/pull/34858#discussion_r3057648646) and can be safely removed after updating the only test where this is load-bearing (b3c9bd7f2df230525c8e339394a315a2c500055d) (see: https://github.com/bitcoin/bitcoin/pull/34858#discussion_r3091085328). The removal also tightens the `addrman_tests/addrman_evictionworks` test to sit exactly on the `ADDRMAN_REPLACEMENT` boundary (4h), catching mutations such as:

```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index d3dae59ae7..d0929c62cb 100644
--- a/src/addrman.cpp
+++ b/src/addrman.cpp
@@ -920,3 +920,3 @@ void AddrManImpl::ResolveCollisions_()
                 // Has successfully connected in last X hours
-                if (current_time - info_old.m_last_success < ADDRMAN_REPLACEMENT) {
+                if (current_time - info_old.m_last_success <= ADDRMAN_REPLACEMENT) {
                     erase_collision = true;
```

The last follow-up item is updating `NodeClockContext` to `FakeNodeClock` to make it clear it is intended for testing (motivated by https://github.com/bitcoin/bitcoin/pull/34858#pullrequestreview-4082110904 and supported in https://github.com/bitcoin/bitcoin/pull/34858#issuecomment-4214352770).

### CODE REVIEW

File: src/test/denialofservice_tests.cpp [hunk 1]
```diff
@@ -252,6 +252,7 @@ BOOST_FIXTURE_TEST_CASE(block_relay_only_eviction, OutboundTest)
     options.m_max_automatic_connections = DEFAULT_MAX_PEER_CONNECTIONS;
 
     connman->Init(options);
+    NodeClockContext clock_ctx{}; // before peer creation so m_connected is deterministic
```
Comment (maflcko) [inline src/test/denialofservice_tests.cpp:255]: nit in the first commit: Maybe  move this after `NodeId id{0};` in the second line? This makes it clearer, because readers and reviewers don't have to think about why this test case is mixing real time and fake time for the initial part.

File: src/test/addrman_tests.cpp [hunk 2]
```diff
@@ -98,7 +98,7 @@ BOOST_AUTO_TEST_CASE(addrman_simple)
 BOOST_AUTO_TEST_CASE(addrman_terrible_many_failures)
 {
     auto now = Now<NodeSeconds>();
-    SetMockTime(now - (ADDRMAN_MIN_FAIL + 24h));
+    NodeClockContext clock_ctx{now - (ADDRMAN_MIN_FAIL + 24h)};
```
Comment (maflcko) [inline src/test/addrman_tests.cpp:101]: nit: Should it not be possible to remove the `now` variable by using:

```
NodeClockContext clock_ctx{};

....

clock_ctx += ADDRMAN_MIN_FAIL + 24h;

```

?

File: src/test/util/time.h [hunk 3]
```diff
@@ -44,8 +44,8 @@ class NodeClockContext
     /// Initialize with the given time.
     explicit NodeClockContext(NodeSeconds init_time) { set(init_time); }
     explicit NodeClockContext(std::chrono::seconds init_time) { set(init_time); }
-    /// Initialize with current time, using the next tick to avoid going back by rounding to seconds.
-    explicit NodeClockContext() { set(++Now<NodeSeconds>().time_since_epoch()); }
+    /// Initialize with current time.
+    explicit NodeClockContext() { set(Now<NodeSeconds>().time_since_epoch()); }
```
Comment (maflcko) [inline src/test/util/time.h:48]: nit: Was thinking that this could go into a more extreme self-checking direction. It could advance the time on constructions by 100h. This way, it should catch bugs, where the context is constructed too late. For example, it may detect the "strictness bug" fixed in the prior commit.

Diff:

```
diff --git a/src/test/util/time.h b/src/test/util/time.h
index 8cd9a1dafc..527014f56b 100644
--- a/src/test/util/time.h
+++ b/src/test/util/time.h
@@ -47,3 +47,3 @@ public:
     /// Initialize with current time, using the next tick to avoid going back by rounding to seconds.
-    explicit NodeClockContext() { set(++Now<NodeSeconds>().time_since_epoch()); }
+    explicit NodeClockContext() { set(100h+Now<NodeSeconds>().time_since_epoch()); }
 
```

Output:

```
$ ./bld-cmake/bin/test_bitcoin -t denialofservice_tests
Running 5 test cases...
test/denialofservice_tests.cpp(101): error: in "denialofservice_tests/outbound_slow_chain_eviction": check !to_send.empty() has failed

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
```

However, this points to another "bug" in `	 27 - chainstate_write_tests (Failed)`, which I presume is due to some leftover init state from `TestingSetup`. I presume it would be possible to fix by providing a `-mocktime=..` startup option in this unit test, or by creating an optional `NodeClockContext` member field inside the `TestingSetup`, which is set in this unit test, but all of this sounds tedious, so up to you, and this is just a nit.

I only wanted to share some more thoughts on this commit, but anything is fine here: Dropping the commit, and leaving this as-is, Increasing the time with some other fixes, or tolerating the time to go backwards.

**Review by maflcko — APPROVED** [review]
lgtm. I just left some nits, but feel free to ignore those

File: src/test/denialofservice_tests.cpp (hunk 1 above)
Comment (seduless) [reply src/test/denialofservice_tests.cpp]: I agree, that does make it easier to reason about. Moved to after `NodeId id{0}` and dropped the comment since the placement is self-documenting now.

File: src/test/addrman_tests.cpp (hunk 2 above)
Comment (seduless) [reply src/test/addrman_tests.cpp]: Nice suggestion, thanks for flagging. I included this change in the latest push and verified that only the relative time gap matters for this test.

File: src/test/util/time.h (hunk 3 above)
Comment (seduless) [reply src/test/util/time.h]: This is a neat idea. I tried both +100h and -100h and they surface different failures. +100h catches the failures you mentioned, but the hoist fix in the first commit only surfaced with -100h. I'd prefer to keep the constructor simple for now since it's not obvious which direction to pick.

**Comment by seduless** [comment]
Thanks for the review! Addressed the placement and addrman feedback and responded to the constructor offset idea in the thread.

**Comment by kevkevinpal** [comment]
ACK [a236a46](https://github.com/bitcoin/bitcoin/pull/35114/commits/a236a4649a02003ce7a374f0f1789c762eb8f7b5)

I reviewed the commits before the scripted diff, and they look good. The scripted diff seems to work well and is a straightforward swap.

---

One nit is that I think it makes more sense to rename `NodeClockContext` to `MockNodeClock` instead of `FakeNodeClock` since we use `SetMockTime` within the class and also some of the comments in the class mention mock time.

The word "Fake" feels off to me. Nonetheless, this is not blocking

**Review by frankomosh — COMMENTED** [review]
tACK a236a4649a02003ce7a374f0f1789c762eb8f7b5. Full test suite runs clean. Also verified the mutation described in the PR description < to <= on `addrman.cpp:921`, is correctly caught by `addrman_evictionworks`

File: src/test/util/time.h (hunk 3 above)
Comment (maflcko) [inline src/test/util/time.h]: nit in 0756e53e13e0e04308ec4bc3683ac927f0c1430e: If you drop the ++operator, the `time_since_epoch` can also be dropped. It was only needed to work around a stdlib bug that I fixed in https://github.com/llvm/llvm-project/commit/a17e97e6778b2cd4114052faf6ee25db330ef405

File: src/test/rpc_tests.cpp [hunk 4]
```diff
@@ -355,7 +355,7 @@ BOOST_AUTO_TEST_CASE(rpc_ban)
     const int64_t ban_duration{o1.find_value("ban_duration").getInt<int64_t>()};
     const int64_t time_remaining{o1.find_value("time_remaining").getInt<int64_t>()};
     BOOST_CHECK_EQUAL(adr.get_str(), "127.0.0.0/24");
-    BOOST_CHECK_EQUAL(banned_until, time_remaining_expected + now.count());
+    BOOST_CHECK_EQUAL(banned_until, time_remaining_expected + Now<NodeSeconds>().time_since_epoch().count());
```
Comment (maflcko) [inline src/test/rpc_tests.cpp]: nit in 99b33f2dc59dd10695d590b59d1727d1b2969e27: Sorry for missing this earlier, but according to the docs of `TicksSinceEpoch` , this should use `TicksSinceEpoch<...>(NodeClock::now())` instead?

Comment (maflcko) [reply src/test/rpc_tests.cpp]: An alternative would be:

```
    BOOST_CHECK_EQUAL(NodeSeconds{banned_until}, time_remaining_expected + Now<NodeSeconds>());
```

(The general idea is to avoid `count()`, where possible)

**Review by maflcko — APPROVED** [review]
everything lgtm, but I guess this needs rebase.

Also, I've seen the context being misused in other pull requests, so maybe it could make sense to enforce there is only one active one at any point in time? Something like this shoudl work:

```diff
diff --git a/src/test/util/time.h b/src/test/util/time.h
index 8cd9a1dafc..bde2258618 100644
--- a/src/test/util/time.h
+++ b/src/test/util/time.h
@@ -10,2 +10,16 @@
 
+/// CRTP Helper to limit a class to at most one at a time.
+template <class T>
+class LimitOne
+{
+public:
+    LimitOne() { Assert(g_T_available) = false; }
+    ~LimitOne() { g_T_available = true; }
+    LimitOne(const LimitOne&) = delete;
+    LimitOne& operator=(const LimitOne&) = delete;
+
+private:
+    static inline bool g_T_available{true};
+};
+
 
@@ -13,3 +27,3 @@
 /// and reset it after use in a test.
-class SteadyClockContext
+class SteadyClockContext : public LimitOne<SteadyClockContext>
 {
@@ -38,3 +52,3 @@ public:
 /// and reset it after use in a test.
-class NodeClockContext
+class NodeClockContext : public LimitOne<NodeClockContext>
 {
```

Otherwise:

review ACK a236a4649a02003ce7a374f0f1789c762eb8f7b5 🎯

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK a236a4649a02003ce7a374f0f1789c762eb8f7b5 🎯
VMm6UpXnOozje+1LiXQg76D5mtMSOmMtSV7WTFSVbERMhetZIMF0QG5uWNVq8RE314nPnR1ck5BrPuUIJf9HCw==
```

</details>

**Comment by sedited** [comment]
@seduless can you take a look at the CI failures here?

**Comment by fanquake** [comment]
@seduless are you still working on this?

**Comment by seduless** [comment]
@fanquake Thanks for the ping, I'll rebase and work through the open comments.

**Comment by seduless** [comment]
Rebased and addressed all open comments, ready for another round of review.

re: https://github.com/bitcoin/bitcoin/pull/35114#issuecomment-4343799723

> One nit is that I think it makes more sense to rename `NodeClockContext` to `MockNodeClock` instead of `FakeNodeClock` since we use `SetMockTime` within the class and also some of the comments in the class mention mock time.

Thanks for the review! Since no one feels strongly about the rename I'd prefer to leave it as-is. Happy to update though if others feel strongly.

re: https://github.com/bitcoin/bitcoin/pull/35114#pullrequestreview-4303514147

> Also, I've seen the context being misused in other pull requests, so maybe it could make sense to enforce there is only one active one at any point in time? Something like this shoudl work

Thanks for the diff. I agree that's a useful improvement to remove the footgun and I applied the diff with the `FakeNodeClock` name and verified it catches misuse at runtime.

re: https://github.com/bitcoin/bitcoin/pull/35114#issuecomment-4512857441

> @seduless can you take a look at the CI failures here?

CI broke because the `cmpctblock` fuzz harness (#33300) landed on `master` using `NodeClockContext`, which the scripted-diff rename didn't catch since that file didn't exist when this branch was pushed. Updating the name during rebase fixes it.

File: src/test/util/time.h (hunk 3 above)
Comment (seduless) [reply src/test/util/time.h]: Thanks for the context, dropped it.

File: src/test/rpc_tests.cpp (hunk 4 above)
Comment (seduless) [reply src/test/rpc_tests.cpp]: Nice catch, updated to use `TicksSinceEpoch`.

**Comment by maflcko** [comment]
Only change is addressing nits, rebase, and adding a new commit.

re-ACK 35a814a045fb1b90801ab8781db5da7c39099ea3 🛒

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK 35a814a045fb1b90801ab8781db5da7c39099ea3 🛒
9JSPsOX3CkRc76GDj7G17GLJGhnW5z4VfKWCfZIWpvBXnpke5H8x1jr1MynDDe42FPGApkNQFJtH5L0WYFGnBg==
```

</details>

**Review by sedited — APPROVED** [review]
ACK 35a814a045fb1b90801ab8781db5da7c39099ea3

**Comment by maflcko** [comment]
Followed-up for the remaining unit-test calls in https://github.com/bitcoin/bitcoin/pull/35497
